task/control-plane-agent/src/mgs_sidecar.rs

```diff
@@ -166,19 +237,197 @@ impl MgsHandler {
+            return Err(GwMonorailError::TimeIsTooLong);
+        }
+
+        let Some(vlan_index) = VLANS.iter().position(|v| v.vid == vid) else {
+            return Err(GwMonorailError::InvalidVLAN);
+        };
+
+        if VLANS[vlan_index].trusted {
+            return Err(GwMonorailError::AlreadyTrusted);
+        }
+
+        let Some((last_challenge, challenge_time)) =
+            core::mem::take(&mut self.last_challenge)
+        else {
+            ringbuf_entry!(Trace::NoChallenge);
+            return Err(GwMonorailError::UnlockAuthFailed);
+        };
+
+        if challenge != last_challenge {
+            ringbuf_entry!(Trace::WrongChallenge);
+            return Err(GwMonorailError::UnlockAuthFailed);
+        }
+
+        let now = sys_get_timer().now;
+        if now >= challenge_time + CHALLENGE_EXPIRATION_TIME_SECS * 1000 {
+            ringbuf_entry!(Trace::ChallengeExpired);
+            return Err(GwMonorailError::UnlockAuthFailed);
+        }
+
+        // Check that the response is valid for our challenge.  Right now,
+        // there's only one option, but we'll use a match to check for the
+        // possibility of new types being added.
+        match (challenge, response) {
+            (UnlockChallenge::Trivial, UnlockResponse::Trivial) => Ok(()),
+        }?;
+        ringbuf_entry!(Trace::UnlockAuthSucceeded);
+
+        // Pick how long we'll trust things for
+        let now = sys_get_timer().now;
+        let trust_until = now + u64::from(time_sec) * 1000;
+
+        // Reconfigure the management network for arbitrary SP access
+        if let Err(e) = self.monorail.unlock_vlans(trust_until) {
+            ringbuf_entry!(Trace::MonorailUnlockFailed(e));
+            return Err(GwMonorailError::UnlockFailed);
+        }
+
+        // Reconfigure the net task to temporarily trust this port's VLAN
+        let net = task_net_api::Net::from(crate::NET.get_task_id());
+        if let Err(e) = net.trust_vlan(vid, trust_until) {
+            ringbuf_entry!(Trace::TrustVLanFailed(e));
+            return Err(GwMonorailError::UnlockFailed);
+        }
+        ringbuf_entry!(Trace::UnlockedUntil(vid, trust_until));
+
+        // Reconfigure our own internal state to accept messages
+        self.locked[vlan_index] = LockState::UnlockedUntil(trust_until);
+
+        Ok(())
+    }
+
+    fn lock(&mut self, vid: u16) -> Result<(), GwMonorailError> {
+        ringbuf_entry!(Trace::Locking(vid));
+
+        let net = task_net_api::Net::from(crate::NET.get_task_id());
+        net.distrust_vlan(vid).map_err(|e| {
+            ringbuf_entry!(Trace::DistrustVLanFailed(e));
+            GwMonorailError::UnlockFailed
+        })?;
+        self.monorail.lock_vlans().map_err(|e| {
+            ringbuf_entry!(Trace::MonorailUnlockFailed(e));
+            GwMonorailError::UnlockFailed
+        })?;
+
+        let i = VLANS.iter().position(|v| v.vid == vid).unwrap();
+        self.locked[i] = LockState::Locked;
```

Comment:
@jgallagher writes

> If `net.distrust_vlan(_)` succeeds but `monorail.lock_vlans()` fails, should we update our state?

Reply:
For both this and the above question, I've made `control-plane-agent` err on the side of being locked (e.g. it locks itself before trying to calling any RPC functions).

This can still leave the system in a partially-consistent state, but the caller will always get an error indicating that something went wrong.  There's no real way to guarantee a consistent state; attempts to recover could _also_ fail (if we believe the initial call could fail).